### ISSUE TITLE
Factor out BufferedPull(s)

### ISSUE THREAD

**Issue by mijicd** [zio/zio#3962]
Due to historical reasons ZStream "leaks" BufferedPull logic in some combinators. The implementer should find these and make sure BufferedPull is (re)used wherever possible.

cc @msvaljek

**Comment by msvaljek** [comment]
I'll take this one!

**Comment by msvaljek** [comment]
@mijicd sorry I won't manage to do this one, the issue is free for the taking again.

**Comment by mghildiy** [comment]
I am a Java programmer, and have basic understanding of Scala. Can I take it?

**Comment by mghildiy** [comment]
Also, is there a resource where I can read about ZIO. It may help me to work on beginner issues.

**Comment by mijicd** [comment]
@mghildiy yes, feel free to. Reach out to me if you need any help. Regarding introductory materials, there are plenty references on our github page, but I would say these two will bring you up to speed in this particular ticket:

- https://youtu.be/TWdC7DhvD8M
- https://youtu.be/OiuKbpMOKsc

**Comment by mghildiy** [comment]
Cloned the repo and found abstract class ZStream. So for this ticket I need to understand this class and modify it.
Jut confirming it here.

**Comment by mijicd** [comment]
@mghildiy that's correct. Check out the BufferedPull and reolsce duplicated code with appropriate method calls.

**Comment by emilhotkowski** [comment]
Is it up for grab? I could help :)

**Comment by adamgfraser** [comment]
@vigoo Is this still relevant?
